autosklearn/data/feature_validator.py

```diff
@@ -0,0 +1,399 @@
+                which columns should be treated as categorical
+        """
+        # Register the user provided feature types
+        if feat_type is not None:
+            if hasattr(X_train, "iloc"):
+                raise ValueError("When providing a DataFrame to Auto-Sklearn, we extract "
+                                 "the feature types from the DataFrame.dtypes. That is, "
+                                 "providing the option feat_type to the fit method is not "
+                                 "supported when using a Dataframe. Please make sure that the "
+                                 "type of each column in your DataFrame is properly set. "
+                                 "More details about having the correct data type in your "
+                                 "DataFrame can be seen in "
+                                 "https://pandas.pydata.org/pandas-docs/stable/reference"
+                                 "/api/pandas.DataFrame.astype.html")
+            # Some checks if feat_type is provided
+            if len(feat_type) != np.shape(X_train)[1]:
+                raise ValueError('Array feat_type does not have same number of '
+                                 'variables as X has features. %d vs %d.' %
+                                 (len(feat_type), np.shape(X_train)[1]))
+            if not all([isinstance(f, str) for f in feat_type]):
+                raise ValueError('Array feat_type must only contain strings.')
+
+            for ft in feat_type:
+                if ft.lower() not in ['categorical', 'numerical']:
+                    raise ValueError('Only `Categorical` and `Numerical` are '
+                                     'valid feature types, you passed `%s`' % ft)
+
+            # Here we register proactively the feature types for
+            # Processing Numpy arrays
+            self.feature_types = feat_type
+
+        self._check_data(X_train)
+
+        if X_test is not None:
+            self._check_data(X_test)
+
+            if np.shape(X_train)[1] != np.shape(X_test)[1]:
+                raise ValueError("The feature dimensionality of the train and test "
+                                 "data does not match train({}) != test({})".format(
+                                     np.shape(X_train)[1],
+                                     np.shape(X_test)[1]
+                                 ))
+
+    def _check_data(
+        self,
+        X: typing.Union[np.ndarray, pd.DataFrame],
+    ) -> None:
+        """
+        Feature dimensionality and data type checks
+
+        Parameters
+        ----------
+            X: typing.Union[np.ndarray, pd.DataFrame]
+                A set of features that are going to be validated (type and dimensionality
+                checks) and a encoder fitted in the case the data needs encoding
+        """
+
+        if not isinstance(X, (np.ndarray, pd.DataFrame, list)) and not scipy.sparse.issparse(X):
```

Comment:
Sparse scipy arrays aren't part of the type annotation, could you please fix this?

Reply:
Done!